docs/writing/posts/retrieval-is-more-than-pure-chance.md

```diff
@@ -0,0 +1,675 @@
+
+If we have 4 items `a,b,c,d` where only `a` is the relevant item, then our Reciprocal Rank(RR) is simply 1. If `c` is the relevant item,then our RR is going to be `1/3`. But, if all of the items are irrelevant, then our RR is going to be 0
+
+ We can express this in the form of a function as 
+
+```python
+def calculate_mrr(chunk_id:str,predictions:List[str]):
+    return 0 if chunk_id not in predictions else 1 / (predictions.index(chunk_id) + 1)
+```
+
+Once we've performed this calculation for a few different retrievals, then the MRR is simply the mean of the individual RRs for each list of retrieved chuns.
+
+### Calculation
+
+Now, let's bring it all together in a single call 
+
+```python
+results = [["a", "b", "c"], ["b", "a", "c"], ["b", "c", "a"]]
+
+evals = {
+    "MRR":calculate_mrr,
+    "NDCG":calculate_ndcg
+}
+evaluations = [
+    {metric: fn("a", result) for metric, fn in evals.items()} for result in results
+]
+
+df = pd.DataFrame(evaluations)
+console = Console()
+console.print(df)
+```
+
+This gives us the values as seen below
+
+
+|MRR |     NDCG |
+| --- | --- |
+| 1.00 |  1.00 |
+| 0.50 |  0.63 |
+| 0.33 |  0.50 |
+
+
+As we can see, as the relevant item ( in this example above 'a' ) gets shifted further, our MRR and NDCG decrease slowly. But, MRR decreases significantly faster and more aggressively than that of NDCG.
+
+### Varying K
+
+Now, we also want to make sure we can calculate these metrics for different values of `k` - to do so, we can use a simple decorator which passes a slice of the list of elements into the metric calculation. 
+
+```python
+def slice_predictions_decorator(num_elements: int):
+    def decorator(func):
+        def wrapper(chunk_id, predictions):
+            sliced_predictions = predictions[:num_elements]
+            return func(chunk_id, sliced_predictions)
+
+        return wrapper
+
+    return decorator
+```
+
+Let's scale our original example just now to 5 values where `a` gets shifted back progressively by one position on each iteration.
+
+
+```python
+evals = {}
+SIZES = [3, 5]
+for size in SIZES:
+    evals[f"MRR@{size}"] = slice_predictions_decorator(size)(calculate_mrr)
+
+for size in SIZES:
+    evals[f"NDCG@{size}"] = slice_predictions_decorator(size)(calculate_ndcg)
+
+results = [
+    ["a", "b", "c", "d", "e"],
+    ["e", "a", "b", "c", "d"],
+    ["d", "e", "a", "b", "c"],
+    ["c", "d", "e", "a", "b"],
+    ["b", "c", "d", "e", "a"],
+]
+
+evaluations = [
+    {metric: fn("a", result) for metric, fn in evals.items()} for result in results
+]
+
+df = pd.DataFrame(evaluations)
+console = Console()
+console.print(df)
+```
+
+
+This gives us the output as seen below. It makes sense to have a MRR@3/NDCG@3 at 0 from the 4th element onwards since by that point, `a` no longer exists within the slice of the first 3 elements that we're takking.
+
+
+| MRR@3 | MRR@5 | NDCG@3 | NDCG@5 |
```

Comment:
focus on precision and recall,

NDCG makes more snese if we actually have a ranking metric which we don't